### PR TITLE
Backend code

### PR DESCRIPTION
**Pull request by radl97**
This is a version of the backend where the admin role management is
missing. Coming in next patch. Also the file uploads/downloadis

### CODE REVIEW

File: API/main.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    caff = CAFF.query.filter_by(name=file.filename, uploader=current_user.id).first()
+    if caff or file.filename.split('.')[-1] != "caff":
+        return Response(response='File already uploaded', status=400)
+    new_caff = CAFF(url=upload_folder+"\\"+file.filename, name=file.filename, comments =0, uploader=current_user.id)
+    db.session.add(new_caff)
+    db.session.commit()
+    file.save(upload_folder+"\\"+file.filename)
+    return Response(response="Upload succesful", status=200)
+
+@main.route('/caffs/<caff_id>', methods=['GET'])
```
Comment (radl97) [inline API/main.py]:
```suggestion
@main.route('/caffs/<int: caff_id>', methods=['GET'])
```

I've found this here: https://stackoverflow.com/questions/14350920/define-a-route-for-url-ending-with-integer-in-python

File: API/main.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    json_data = {
+        "id" : caff.id,
+        "preview_url" : caff.url, #wtf?
+        "comments" : comments_list,
+        "number_of_comments" : caff.comments,
+        "uploader" : {"id" : uploader.id ,"name" : uploader.name}
+    }
+    return json_data
+
+@main.route('/caffs/<caff_id>', methods=['DELETE'])
```
Comment (radl97) [inline API/main.py]:
```suggestion
@main.route('/caffs/<int: caff_id>', methods=['DELETE'])
```

File: API/main.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    caff = db.session.query(CAFF).get(caff_id)
+    db.session.delete(caff)
+    comments = db.session.query(Comment).all()
+    for comment in comments:
+        if comment.caff_id == int(caff_id):
+            db.session.delete(comment)
+    db.session.commit()
+    return Response(status=200)
+
+@main.route('/caffs/<caff_id>/comments', methods=['POST'])
```
Comment (radl97) [inline API/main.py]:
```suggestion
@main.route('/caffs/<int: caff_id>/comments', methods=['POST'])
```

File: API/main.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+def addComment(caff_id):
+    comment_text = request.json["text"]
+    comment = Comment(caff_id= caff_id, user_id=current_user.id, text=comment_text )
+    db.session.add(comment)
+    caff = db.session.query(CAFF).get(caff_id)
+    caff.comments += 1
+    db.session.commit()
+    return Response(status=200)
+
+@main.route('/caffs/<caff_id>/comments/<comment_id>', methods=['DELETE'])
```
Comment (radl97) [inline API/main.py]:
```suggestion
@main.route('/caffs/<int: caff_id>/comments/<int: comment_id>', methods=['DELETE'])
```

File: API/main.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+@login_required
+def deleteComment(caff_id, comment_id):
+    caff = db.session.query(CAFF).get(caff_id)
+    comment = db.session.query(Comment).get(comment_id)
+    db.session.delete(comment)
+    caff.comments -= 1
+    db.session.commit()
+    return Response(status=200)
+
+@main.route('/caffs/<caff_id>/download', methods=['GET'])
```
Comment (radl97) [inline API/main.py]:
```suggestion
@main.route('/caffs/<int: caff_id>/download', methods=['GET'])
```

File: API/main.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    db.session.commit()
+    return Response(status=200)
+
+@main.route('/caffs/<caff_id>/comments/<comment_id>', methods=['DELETE'])
+@login_required
+def deleteComment(caff_id, comment_id):
+    caff = db.session.query(CAFF).get(caff_id)
+    comment = db.session.query(Comment).get(comment_id)
+    db.session.delete(comment)
+    caff.comments -= 1
```
Comment (radl97) [inline API/main.py]: isn't it `number_of_comments`? Or this is just DB name?

**Comment by radl97** [comment]
Can you commit a requirements.txt please?

File: API/main.py (hunk 6 above)
Comment (CassioAI) [reply API/main.py]: just DB field name:: 
image

File: backend/API/__init__.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
+import os
+from flask import Flask
+from flask_sqlalchemy import SQLAlchemy
+from flask_login import LoginManager
+
+# init SQLAlchemy so we can use it later in our models
+db = SQLAlchemy()
+upload_folder ="uploads\\"
```
Comment (radl97) [inline backend/API/__init__.py]: this kills linux versions (I now have a `uploads\` directory lol)

```suggestion
upload_folder ="uploads" +  os.pathsep
```

File: backend/API/main.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+def upload():
+    #prolly works
+    file= request.files['file']
+    caff = CAFF.query.filter_by(name=file.filename, uploader=current_user.id).first()
+    if caff or file.filename.split('.')[-1] != "caff":
+        return Response(response='File already uploaded', status=400)
+    new_caff = CAFF(url=upload_folder+"\\"+file.filename, name=file.filename, comments =0, uploader=current_user.id)
+    db.session.add(new_caff)
+    db.session.commit()
+    file.save(upload_folder+"\\"+file.filename)
```
Comment (radl97) [inline backend/API/main.py]: upload_folder already contains the path separator (`\`)

```suggestion
    file.save(upload_folder+file.filename)
```

File: backend/API/main.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+@main.route('/caffs', methods=['POST'])
+@login_required
+def upload():
+    #prolly works
+    file= request.files['file']
+    caff = CAFF.query.filter_by(name=file.filename, uploader=current_user.id).first()
+    if caff or file.filename.split('.')[-1] != "caff":
+        return Response(response='File already uploaded', status=400)
+    new_caff = CAFF(url=upload_folder+"\\"+file.filename, name=file.filename, comments =0, uploader=current_user.id)
```
Comment (radl97) [inline backend/API/main.py]: upload_folder already contains the path separator (`\`)

```suggestion
    new_caff = CAFF(url=upload_folder+file.filename, name=file.filename, comments =0, uploader=current_user.id)
```

**Pull request merged**